src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEMethodSymbol.cs

```diff
@@ -193,10 +193,16 @@ private UncommonFields CreateUncommonFields()
                 retVal._lazyObsoleteAttributeData = ObsoleteAttributeData.Uninitialized;
             }
 
-            if (!_packedFlags.IsUseSiteDiagnosticPopulated)
-            {
-                retVal._lazyUseSiteDiagnostic = CSDiagnosticInfo.EmptyErrorInfo; // Indicates unknown state.
-            }
+            //
+            // Do not set _lazyUseSiteDiagnostic !!!!
+            //
+            // "null" Indicates "no errors" or "unknown state".
+            // and we know which one of the states we have from IsUseSiteDiagnosticPopulated
+            //
+            // setting _lazyUseSiteDiagnostic to a sentinel value here would introduce
+            // a number of states with various permutations between IsUseSiteDiagnosticPopulated, UncommonFields and _lazyUseSiteDiagnostic
```

Comment:
One of possible race scenarios:

thread `#1` sees that IsUseSiteDiagnosticPopulated is not set, goes to compute and publish diagnostics
thread `#2` sees that IsUseSiteDiagnosticPopulated is not set, goes to compute and publish diagnostics
thread `#1` finds no issues and sets IsUseSiteDiagnosticPopulated
thread `#3` does something else and indirectly creates UncommonFields with a sentinel value for the use site diagnostics
thread `#2` checks again IsUseSiteDiagnosticPopulated and since it is set (by thread #1) happily returns the sentinel as the actual value!!!

---
In reply to: [124587823](https://github.com/dotnet/roslyn/pull/20497#discussion_r124587823) [](ancestors = 124587823)